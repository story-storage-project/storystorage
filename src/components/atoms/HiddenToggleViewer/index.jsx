import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const Wrapper = styled.span(
  ({ ...props }) => `
    display: ${props.reverse ? 'none' : 'visible'};
    @media ${props.theme.viewSize[props.hiddenView]} {
      display: ${
        (props.reverse && 'block') || (props.toggle ? 'visible' : 'none')
      };
    }
`,
);

export default function HiddenToggleViewer({ children, ...props }) {
  return (
    <Wrapper aria-label="component" {...props}>
      {children}
    </Wrapper>
  );
}

HiddenToggleViewer.propTypes = {
  children: PropTypes.node.isRequired,
};
