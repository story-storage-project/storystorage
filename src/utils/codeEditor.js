export function getCaretPosition(textBox, caretPosition) {
  if (textBox.setSelectionRange) {
    textBox.focus();
    textBox.setSelectionRange(caretPosition, caretPosition);
  }

  return caretPosition;
}

export function setCaretPosition(textBox, pos) {
  if (textBox.setSelectionRange) {
    textBox.focus();
    textBox.setSelectionRange(pos, pos);
  }
}

export function insertTab(event, textBox, caretPositionArg) {
  event.preventDefault();
  const caretPosition = getCaretPosition(textBox, caretPositionArg);
  const preText = textBox.value.substring(0, caretPosition);
  const postText = textBox.value.substring(caretPosition, textBox.value.length);
  const copyTextBox = textBox;
  copyTextBox.value = `${preText}\t${postText}`;

  setCaretPosition(textBox, caretPosition + 1);

  return copyTextBox.value;
}

export function insertText(textBox, caretPositionArg, text) {
  const caretPosition = getCaretPosition(textBox, caretPositionArg);
  const preText = textBox.value.substring(0, caretPosition);
  const postText = textBox.value.substring(caretPosition, textBox.value.length);
  const copyTextBox = textBox;
  copyTextBox.value = `${preText}${text}${postText}`;

  setCaretPosition(textBox, caretPosition);

  return copyTextBox.value;
}
