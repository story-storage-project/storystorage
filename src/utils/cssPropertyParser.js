import htmlElementsSet from '../data/htmlElements';
import {
  cssKeyOfcssProperties,
  jsxKeyOfCssProperties,
} from '../data/cssJsxProperties';
import { VALIDATION_ERROR_MESSAGE } from '../constants/errorMessage';

function validateCss(css) {
  if (!css.length) {
    return VALIDATION_ERROR_MESSAGE.NULL;
  }

  const validateBrackets = css.match(/{/g).length === css.match(/}/g).length;
  const validateSemicolon = css.match(/:/g).length === css.match(/;/g).length;

  if (validateBrackets && validateSemicolon) {
    return 'pass';
  }

  if (!validateBrackets && !validateSemicolon) {
    return VALIDATION_ERROR_MESSAGE.CSS.MISSING_ALL;
  }

  if (!validateBrackets) {
    return VALIDATION_ERROR_MESSAGE.CSS.MISSING_BRACKETS;
  }

  return VALIDATION_ERROR_MESSAGE.CSS.MISSING_SEMICOLON;
}

function getCssMode(key) {
  const firstCha = key[0];
  if (firstCha === '.') {
    return 'class';
  }

  if (firstCha === '#') {
    return 'id';
  }

  if (!htmlElementsSet.has(key)) {
    return VALIDATION_ERROR_MESSAGE.CSS.INCORRECT_KEY;
  }

  return 'element';
}

function getCssAllKey(css) {
  let key = '';
  let isInsideBracket = false;

  const cssAllKey = [...css].reduce((acc, curr) => {
    if (curr === '{') {
      isInsideBracket = true;

      if (key.includes(',')) {
        const sliceKeyArr = key.split(',');
        const validatedKeyArr = [];

        sliceKeyArr.forEach(sliceKey => {
          const mode = getCssMode(sliceKey);
          const keyObject = {};

          if (mode !== VALIDATION_ERROR_MESSAGE.CSS.INCORRECT_KEY) {
            keyObject.mode = mode;
            keyObject.key = sliceKey;

            validatedKeyArr.push(keyObject);
          }
        });
        acc.push(validatedKeyArr);
        key = '';

        return acc;
      }

      const mode = getCssMode(key);

      if (mode === VALIDATION_ERROR_MESSAGE.CSS.INCORRECT_KEY) {
        return acc;
      }

      const keyObject = { mode, key };

      if (key.length >= 1) {
        acc.push(keyObject);
      }

      key = '';

      return acc;
    }

    if (curr === '}') {
      isInsideBracket = false;
      return acc;
    }

    if (!isInsideBracket) {
      key += curr;
    }

    return acc;
  }, []);

  return cssAllKey;
}

function removeSpace(string) {
  return string.replace(/\s|"|'/g, '');
}

function getCssAllValue(css) {
  let isInsideBracket = false;
  let hasProperty = false;
  let property = '';
  let value = '';
  let obj = {};

  const cssAllValue = [...css].reduce((acc, curr) => {
    if (curr === ';') {
      const prop = removeSpace(property);

      if (!(jsxKeyOfCssProperties[prop] || cssKeyOfcssProperties[prop])) {
        return acc;
      }

      obj[prop] = value;

      hasProperty = false;
      property = '';
      value = '';
      return acc;
    }

    if (curr === '{') {
      isInsideBracket = true;
      return acc;
    }

    if (curr === '}') {
      acc.push(obj);
      obj = {};

      isInsideBracket = false;
      return acc;
    }

    if (curr === ':') {
      hasProperty = true;
      return acc;
    }

    if (isInsideBracket) {
      if (!hasProperty) {
        property += curr;
      } else {
        value += curr;
      }
    }

    return acc;
  }, []);

  return cssAllValue;
}

export default function getParsingCss(css) {
  const validateMessage = validateCss(css);

  if (validateMessage !== 'pass') return validateMessage;

  const cssKeyArr = getCssAllKey(removeSpace(css));
  const cssValuesArr = getCssAllValue(css);

  const storedStyleObj = {
    class: {},
    id: {},
    element: {},
  };

  const storeStyle = (keyArr, duplicateValueIndex) => {
    return keyArr.reduce((store, keyObj, index) => {
      if (Array.isArray(keyObj)) {
        return storeStyle(keyObj, index);
      }

      const { mode, key } = keyObj;

      const property = duplicateValueIndex
        ? cssValuesArr[duplicateValueIndex]
        : cssValuesArr[index];

      if (store[mode][key]) {
        Object.assign(property, store[mode][key]);
      }

      // eslint-disable-next-line no-param-reassign
      store[mode][key] = property;

      return store;
    }, storedStyleObj);
  };

  return storeStyle(cssKeyArr);
}
