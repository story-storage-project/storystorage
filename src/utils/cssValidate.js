import { VALIDATION_ERROR_MESSAGE } from '../constants/errorMessage';

export default function validateCss(css) {
  if (!css.length) {
    return VALIDATION_ERROR_MESSAGE.NULL;
  }
  if (
    !css.includes('{') ||
    !css.includes('}') ||
    !css.includes(':') ||
    !css.includes(';')
  ) {
    return VALIDATION_ERROR_MESSAGE.CSS.MISSING_ALL;
  }

  const validateBrackets = css.match(/{/g).length === css.match(/}/g).length;
  const validateSemicolon = true;

  if (validateBrackets && validateSemicolon) {
    return '';
  }

  if (!validateBrackets && !validateSemicolon) {
    return VALIDATION_ERROR_MESSAGE.CSS.MISSING_ALL;
  }

  if (!validateBrackets) {
    return VALIDATION_ERROR_MESSAGE.CSS.MISSING_BRACKETS;
  }

  return VALIDATION_ERROR_MESSAGE.CSS.MISSING_SEMICOLON;
}
