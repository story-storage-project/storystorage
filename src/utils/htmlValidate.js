import { VALIDATION_ERROR_MESSAGE } from '../constants/errorMessage';
import {
  htmlKeyOfhtmlJsxAttributes,
  jsxKeyOfhtmlJsxAttributes,
} from '../data/htmlJsxAttributes';
import htmlElementsSet from '../data/htmlElements';
import keyMaker from './keyMaker';

function basicValidateHtml(html) {
  if (!html.length) {
    return VALIDATION_ERROR_MESSAGE.NULL;
  }

  if (!html.includes('<') || !html.includes('>')) {
    return VALIDATION_ERROR_MESSAGE.HTML.MISSING_BRACKETS;
  }

  const arrowRemoveHtml = html.replace(/=>/g, '');
  const validateBrackets =
    arrowRemoveHtml?.match(/</g)?.length ===
    arrowRemoveHtml?.match(/>/g)?.length;

  return validateBrackets ? '' : VALIDATION_ERROR_MESSAGE.HTML.MISSING_BRACKETS;
}

export function validateHtml(html) {
  if (!html) return;

  const basicValidationError = basicValidateHtml(html);

  if (basicValidationError) {
    return basicValidationError;
  }

  const bracketSplitCode = html.split(/(<!--| |-->|<|>)/);
  let tagName = '';
  let errorMessage = '';

  bracketSplitCode.forEach((item, i) => {
    if (item === '<' && bracketSplitCode[i + 1] !== '/') {
      tagName = bracketSplitCode[i + 1];

      if (tagName[0] === '/') return;

      if (!htmlElementsSet.has(tagName)) {
        errorMessage = VALIDATION_ERROR_MESSAGE.HTML.INCORRECT_ELEMENT;
      }
    }

    if (item === '>') {
      tagName = '';
    }
  });

  return errorMessage;
}

export function exportInlineStyle(code, mode) {
  const splitHtml = code.split(/(<!--| |-->|<|>|\/)/);

  let tagName = '';
  let isStyle = false;
  let style = '';
  const styleStorage = [];

  const addDatasetCode = splitHtml.reduce((acc, item, i) => {
    if (item === '<') {
      tagName = splitHtml[i + 1];

      acc.push(item);
      return acc;
    }

    if (splitHtml[i - 1] === '<') {
      if (item.slice(0, 1) === '/') {
        tagName = '';
        acc.push(item);

        return acc;
      }
      acc.push(item);

      return acc;
    }

    if (isStyle) {
      if (
        mode === 'HTML' &&
        (item.slice(-1) === "'" || item.slice(-1) === '"')
      ) {
        isStyle = false;

        style += item;
        const splitProperty = style.split(/(=)/);
        const newClass = keyMaker(5);

        acc.push(`data-${newClass}=inline `);

        styleStorage.push(
          `.${newClass}{${splitProperty[2].replace(/'|"/g, '')}}`,
        );

        return acc;
      }
      if (mode === 'JSX' && item.slice(-1) === '}') {
        isStyle = false;

        style += item;
        const splitProperty = style.split(/(=)/);
        const removeQuoteAndSemicolon = splitProperty[2]
          .replace(/"|'/g, '')
          .replace(/,/g, ';')
          .replace(/{{/g, '{')
          .replace(/}}/g, '}');

        const newClass = keyMaker(5);

        acc.push(`data-${newClass}=inline `);

        if (removeQuoteAndSemicolon.slice(-2, -1) !== ';') {
          const newCss = `${removeQuoteAndSemicolon.slice(0, -1)};}`;
          styleStorage.push(`.${newClass}${newCss}`);

          return acc;
        }

        styleStorage.push(`.${newClass}${removeQuoteAndSemicolon}`);

        return acc;
      }

      style += item;

      return acc;
    }

    if (tagName && item !== '>') {
      const splitProperty = item.split(/(=)/);

      if (splitProperty[0].length > 1) {
        if (splitProperty[0] === 'style') {
          if (
            (mode === 'HTML' && splitProperty[2].slice(-1) === "'") ||
            splitProperty[2].slice(-1) === '"'
          ) {
            const newClass = keyMaker(5);

            acc.push(`data-${newClass}=inline `);
            styleStorage.push(
              `.${newClass}{${splitProperty[2].replace(/'|"/g, '')}}`,
            );

            return acc;
          }
          if (mode === 'JSX' && splitProperty[2].slice(-1) === '}') {
            const removeQuoteAndSemicolon = splitProperty[2]
              .replace(/"|'/g, '')
              .replace(/,/g, ';')
              .replace(/{{/g, '{')
              .replace(/}}/g, '}');

            const newClass = keyMaker(5);
            acc.push(`data-${newClass}=inline `);

            if (removeQuoteAndSemicolon.slice(-2, -1) !== ';') {
              const newCss = `${removeQuoteAndSemicolon.slice(0, -1)};}`;
              styleStorage.push(`.${newClass}${newCss}`);

              return acc;
            }

            styleStorage.push(`.${newClass}${removeQuoteAndSemicolon}`);

            return acc;
          }

          isStyle = true;
          style = splitProperty.join('');

          return acc;
        }
        acc.push(item);

        return acc;
      }
      acc.push(item);

      return acc;
    }

    if (item === '>') {
      tagName = '';
      acc.push(item);

      return acc;
    }
    acc.push(item);

    return acc;
  }, []);

  return [addDatasetCode.join(''), styleStorage];
}

export function convertJsxToHtml(html) {
  const splitHtml = html.split(/(<!--| |-->|<|>|\/)/);

  let tagName = '';
  let isStyle = false;

  const convert = splitHtml.reduce((acc, item, i) => {
    if (item === '<!--' || item === '-->') {
      acc.push(item);

      return acc;
    }
    if (item === '<') {
      tagName = splitHtml[i + 1];

      acc.push(item);

      return acc;
    }

    if (item === '/>') {
      acc.push(`></${tagName}>`);

      return acc;
    }

    if (splitHtml[i - 1] === '<') {
      if (item.slice(0, 1) === '/') {
        tagName = '';

        acc.push(item);

        return acc;
      }

      acc.push(item);

      return acc;
    }

    if (isStyle) {
      if (item.slice(-1) === '}') {
        isStyle = false;

        return acc;
      }

      return acc;
    }

    if (tagName && item !== '>') {
      const splitProperty = item.split(/(=)/);

      if (splitProperty[0].length > 1) {
        if (splitProperty[0] === 'style') {
          return acc;
        }

        const convertProperty = jsxKeyOfhtmlJsxAttributes[splitProperty[0]];

        splitProperty[0] = convertProperty || splitProperty[0];

        acc.push(splitProperty.join(''));

        return acc;
      }

      acc.push(item);

      return acc;
    }

    if (item === '>') {
      tagName = '';

      acc.push(item);

      return acc;
    }

    acc.push(item);

    return acc;
  }, []);

  return convert.join('');
}

export function convertHtmlToJsx(Jsx) {
  const splitHtml = Jsx.split(/(<!--| |-->|<|>|\/)/);

  let tagName = '';
  let isStyle = false;

  const convert = splitHtml.reduce((acc, item, i) => {
    if (item === '<!--' || item === '-->') {
      acc.push(item);

      return acc;
    }
    if (item === '<') {
      tagName = splitHtml[i + 1];

      acc.push(item);

      return acc;
    }

    if (item === '/>') {
      acc.push(`></${tagName}>`);

      return acc;
    }

    if (splitHtml[i - 1] === '<') {
      if (item.slice(0, 1) === '/') {
        tagName = '';

        acc.push(item);

        return acc;
      }

      acc.push(item);

      return acc;
    }

    if (isStyle) {
      if (item.slice(-1) === '}') {
        isStyle = false;

        return acc;
      }

      return acc;
    }

    if (tagName && item !== '>') {
      const splitProperty = item.split(/(=)/);

      if (splitProperty[0].length > 1) {
        if (splitProperty[0] === 'style') {
          return acc;
        }

        const convertProperty = htmlKeyOfhtmlJsxAttributes[splitProperty[0]];
        splitProperty[0] = convertProperty || splitProperty[0];

        acc.push(splitProperty.join(''));

        return acc;
      }

      acc.push(item);

      return acc;
    }

    if (item === '>') {
      tagName = '';

      acc.push(item);

      return acc;
    }

    acc.push(item);

    return acc;
  }, []);

  return convert.join('');
}
