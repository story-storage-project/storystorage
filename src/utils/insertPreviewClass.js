export default function insertClass(toBeAddedClass, css) {
  const newClass = `.a${toBeAddedClass}`;
  const splitCss = css.trim().replace(/\n/g, '').split(/({|})/);

  const addPreviewClass = [];
  splitCss.forEach((item, i) => {
    if (i === 0) {
      return addPreviewClass.push(newClass, ' ', item.trim());
    }

    if (item === '}' && i !== splitCss.length - 2) {
      return addPreviewClass.push(item.trim(), newClass, ' ');
    }

    addPreviewClass.push(item.trim());
  });

  return addPreviewClass.join('');
}
