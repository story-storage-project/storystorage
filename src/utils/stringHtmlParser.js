import { DOCUMENT_FRAGMENT_NODE } from "../constants/nodeTypes";

export default function stringHtmlParser(source) {
  const parser = new DOMParser();
  const dom = parser.parseFromString(
    `<!doctype html><body><template>${source}</template></body>`,
    "text/html"
  );

  const nodes = dom.body.firstChild.content.childNodes;

  if (nodes.length > 1) {
    if (nodes[0].tagName !== "REACT.FRAGMENT" && nodes[0].nodeValue !== "<>") {
      alert("안돼");
    }
  }

  return nodes.length === 1
    ? nodes[0]
    : { nodeType: DOCUMENT_FRAGMENT_NODE, childNodes: nodes };
}

// const parse = (source) => {
//   const dom = htmlParser(source);

//   transform2(dom);
//   return dom;
// };
