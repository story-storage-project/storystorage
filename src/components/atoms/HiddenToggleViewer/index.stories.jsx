import { INITIAL_VIEWPORTS } from '@storybook/addon-viewport';
import { node } from 'prop-types';
import React from 'react';
import HiddenToggleViewer from '.';
import ImageIcon from '../ImageIcon';

export default {
  title: 'Atom/HiddenToggleViewer',
  component: HiddenToggleViewer,
  argTypes: {
    children: {
      type: node,
    },
    hiddenView: {
      type: 'string',
      defaultValue: 'tablet',
    },
  },
  parameters: {
    viewport: {
      viewports: INITIAL_VIEWPORTS,
      defaultViewport: 'responsive',
    },
  },
  subcomponents: { ImageIcon },
};

function Template(args) {
  return (
    <HiddenToggleViewer {...args}>
      <ImageIcon icon="menu" alt="menu-icon" pointers />
    </HiddenToggleViewer>
  );
}
export const LabtopViewPort = Template.bind({});
LabtopViewPort.parameters = {
  viewport: {
    defaultViewport: 'tablet',
  },
};

export const TabletViewPort = Template.bind({});
TabletViewPort.parameters = {
  viewport: {
    defaultViewport: 'pixelxl',
  },
};
