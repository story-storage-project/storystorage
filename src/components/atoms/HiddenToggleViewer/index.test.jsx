/* eslint-disable import/no-unresolved */
import React from 'react';
import { screen } from '@testing-library/react';
import 'jest-styled-components';
import { render } from 'test/utils/testUtil';
import base from 'components/theme/default';
import HiddenToggleViewer from '.';

let tree;

describe('HiddenToggleViewer - props.reverse', () => {
  beforeEach(() => {
    tree = render(
      <HiddenToggleViewer hiddenView="tablet" reverse>
        <div>test</div>
      </HiddenToggleViewer>,
    );
  });

  it('should render without crashing', () => {
    expect(tree).toBeDefined();
  });

  it('should be visible a span', () => {
    const span = screen.getByLabelText('component');

    expect(span).toBeInTheDocument();
  });

  it('should be display property is none', () => {
    expect(tree.container.firstChild).toMatchSnapshot();
    expect(tree.container.firstChild).toHaveStyle('display: none');
  });

  it('should be display property is block for the specified viewsize', () => {
    expect(tree.container.firstChild).toMatchSnapshot();
    expect(tree.container.firstChild).toHaveStyleRule('display', 'block', {
      media: base.viewSize.tablet,
    });
  });
});

describe('HiddenToggleViewer - props.toggle', () => {
  it('if toggle is false for the specified viewsize, should be display property is none', () => {
    tree = render(
      <HiddenToggleViewer hiddenView="tablet" toggle={false}>
        <div>test</div>
      </HiddenToggleViewer>,
    );

    expect(tree.container.firstChild).toHaveStyleRule('display', 'none', {
      media: base.viewSize.tablet,
    });
  });

  it('if toggle is true for the specified viewsize, should be display property is visible', () => {
    tree = render(
      <HiddenToggleViewer hiddenView="tablet" toggle>
        <div>test</div>
      </HiddenToggleViewer>,
    );
    expect(tree.container.firstChild).toHaveStyle('display: visible');
    expect(tree.container.firstChild).toHaveStyleRule('display', 'visible', {
      media: base.viewSize.tablet,
    });
  });
});
