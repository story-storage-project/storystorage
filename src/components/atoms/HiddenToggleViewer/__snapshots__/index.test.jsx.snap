// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HiddenToggleViewer - props.reverse should be display property is block for the specified viewsize 1`] = `
.c0 {
  display: none;
}

@media (max-width:900px) {
  .c0 {
    display: block;
  }
}

<span
  aria-label="component"
  class="c0"
>
  <div>
    test
  </div>
</span>
`;

exports[`HiddenToggleViewer - props.reverse should be display property is none 1`] = `
.c0 {
  display: none;
}

@media (max-width:900px) {
  .c0 {
    display: block;
  }
}

<span
  aria-label="component"
  class="c0"
>
  <div>
    test
  </div>
</span>
`;

exports[`HiddenToggleViewer - props.toggle should be display property is block for the specified viewsize 1`] = `
.c0 {
  display: visible;
}

@media (max-width:900px) {
  .c0 {
    display: visible;
  }
}

<span
  aria-label="component"
  class="c0"
>
  <div>
    test
  </div>
</span>
`;

exports[`HiddenToggleViewer - props.toggle should be display property is none 1`] = `
.c0 {
  display: visible;
}

@media (max-width:900px) {
  .c0 {
    display: visible;
  }
}

<span
  aria-label="component"
  class="c0"
>
  <div>
    test
  </div>
</span>
`;
